### ISSUE TITLE
Not all output is being displayed

### ISSUE THREAD

**Issue by Globegitter** [shelljs/shelljs#165]
I don't know why exactly, but if I run an async `exec` command `fig up`, from https://github.com/docker/fig it does not display all of the output.
Don't know how I can help with any more detail, but it is when it logs the output from the docker containers.
If there is any other way I can help nailing down the specifics of this bug, let me know.
(silent is false of course)

**Comment by samartioli** [comment]
I see this also... From time to time I will see output. Take the following example from my gulpfile:

    gulp.task('build', function() {
        var newIteration = ("0000" + (parseInt(build.iteration) + 1)).slice(-4); 
        var newVersion = pkg.version + '-' + newIteration;
        sh.exec('docker build -t my-api:' + newVersion + ' .');
        sh.exec('docker tag -f my-api:' + newVersion + ' my-api:latest');
        build.iteration = newIteration;
        fs.writeFileSync('./build.json', JSON.stringify(build, null, 4));
        sh.exec('docker images my-api');
    });

The first 2 sh.exec's will output, but the:
    
    sh.exec('docker images my-api');

Will sometimes show and sometimes not. 

Race condition?

I have found that as a work around, I can add a callback. That seems to be working.

Example, the last exec becomes:

    sh.exec('docker images my-api', function(code, output) { console.log("Finished");});

**Comment by jakub-g** [comment]
I've noticed similar for `git clone` command on Windows 7 machine.

Consider following js

```js
require('shelljs/global');
exec('git clone https://github.com/jakub-g/empty-repo.git');
```

The result is that I see **only the first line**

```
me@mymachine ~/git
$ node shelltest.js
Cloning into 'empty-repo'...
```

**Comment by jakub-g** [comment]
I checked the implementation of `exec` and it's pretty complicated
https://github.com/shelljs/shelljs/blob/master/src/exec.js

I am interested more about **sync exec**, than async, to output everything, so here are my findings:

nodejs 0.12 comes with `childProcess.execSync` (it's not available in 0.10)

By default the behavior is the same as using shelljs `exec`, i.e. only the first line gets displayed:

```
me@mymachine ~
# require('child_process').execSync('git clone https://github.com/jakub-g/empty-repo.git');
$ node execSyncTest.js 
Cloning into 'empty-repo'...
```

Now things are getting interesting. According to node docs, no `stdio` param passed means the default `pipe` will be used, but it's not the case. When I pass `pipe` I don't see any output:

```
me@mymachine ~
# require('child_process').execSync('git clone https://github.com/jakub-g/empty-repo.git', { stdio: 'pipe' });
$ node execSyncTest.js 
```

**When I pass `inherit` I see all output:**

```
me@mymachine ~
# require('child_process').execSync('git clone https://github.com/jakub-g/empty-repo.git', { stdio: 'inherit' });
$ node execSyncTest.js 
Cloning into 'empty-repo'...
remote: Counting objects: 4, done.
remote: Total 4 (delta 0), reused 0 (delta 0), pack-reused 4
Unpacking objects: 100% (4/4), done.
Checking connectivity... done.
```

Note again `execSync` is only available in node 0.12.

Shelljs uses `execOptions.stdio = [0, 1, 2]` which is equivalent to `'inherit'` but then it creates an adhoc script and does some magic stdout redirections so I don't really know how to enable the fix for shelljs. I would stick to using `childProcess.execSync`.

**Comment by TimothyGu** [comment]
I don't think this is a problem with shell.js. Git does some sketchy tricks with detecting if the stderr stream is a terminal or not. Try this:

```
$ git clone https://github.com/jakub-g/empty-repo.git 2>&1 | cat
Cloning into 'empty-repo'...
$
```

Git only seems to print the first line if the stderr is connected to a terminal, which in this case is not true (stderr → stdout → cat).

I haven't checked docker but I suspect it is a similar case.

**Comment by TimothyGu** [comment]
Indeed, in git-clone(1) it is written as follows:

```
       --progress
           Progress status is reported on the standard error stream by default
           when it is attached to a terminal, unless -q is specified. This
           flag forces progress status even if the standard error stream is
           not directed to a terminal.
```

And as expected:

```
$ git clone --progress https://github.com/jakub-g/empty-repo.git 2>&1 | cat
Cloning into 'empty-repo'...
remote: Counting objects: 4, done.        
remote: Total 4 (delta 0), reused 0 (delta 0), pack-reused 4        
Checking connectivity... done.
$ rm -rf empty-repo
$ node -e "require('child_process').execSync('git clone --progress https://github.com/jakub-g/empty-repo.git');"
Cloning into 'empty-repo'...
remote: Counting objects: 4, done.        
remote: Total 4 (delta 0), reused 0 (delta 0), pack-reused 4        
Checking connectivity... done.
```

**Comment by TimothyGu** [comment]
> Now things are getting interesting. According to node docs, no stdio param passed means the default `pipe` will be used, but it's not the case. When I pass pipe I don't see any output:

Multiple wrong things here:

1. `pipe` is the default for spawnSync, which execSync internally uses, but `pipe` does not mean piping to the stdio streams of the parent:

    > Create a pipe between the child process and the parent process. The parent end of the pipe is exposed to the parent as a property on the `child_process` object as `ChildProcess.stdio[fd]`. Pipes created for fds 0 - 2 are also available as ChildProcess.stdin, ChildProcess.stdout and ChildProcess.stderr, respectively.
    > (https://nodejs.org/api/child_process.html#child_process_options_stdio)

    In contrast, `inherit` does exactly what you want, which is making the child process inherit the streams of the parent.
2. execSync's default is not the same as the `'pipe'`

    > - `stdio` `Array` Child's stdio configuration. (Default: 'pipe')
    >   - **`stderr` by default will be output to the parent process' stderr unless stdio is specified**
    >
    > (https://nodejs.org/api/child_process.html#child_process_child_process_execsync_command_options)

    So basically the default *behavior* is the same as `'pipe'`, but if nothing is specified it would inherit the stderr stream as well. I also agree that the point of `'pipe'` for execSync is moot since the only way the pipes are accessible is through the ChildProcess object corresponding to the process, and yet execSync does not provide access to the ChildProces object.

Hope that this makes a little more sense.

**Comment by jakub-g** [comment]
@TimothyGu thanks for your explanations.

Regarding

>    pipe is the default for spawnSync, which execSync internally uses, but pipe does not mean piping to the stdio streams of the parent:

>>    Create a pipe between the child process and the parent process. The parent end of the pipe is exposed to the parent as a property on the child_process object as ChildProcess.stdio[fd]. Pipes created for fds 0 - 2 are also available as ChildProcess.stdin, ChildProcess.stdout and ChildProcess.stderr, respectively.
    (https://nodejs.org/api/child_process.html#child_process_options_stdio)

It's a bit confusing to me. Should I understand it as follows:

> `pipe` means there's **a new pipe created**; the child pipes stuff and to the parent, and it's up to parent to decide what to do with this stuff (output it to its own stdout etc.). While `inherit` means there are no new pipes created, the existing stdin, stdout, stderr are reused, so it looks like the child process was a kind of "subroutine" of the parent process.

Is this the case?

I've noticed there are some other differences in `shelljs/exec` and `child_process.execSync` API, shelljs's exec more resembles bash (checking the error code), while for execSync you have to `try-catch`. Need to make up my mind which I like more.

**Comment by TimothyGu** [comment]
> Is this the case?

Exactly.

**Comment by ariporad** [comment]
@TimothyGu: Thanks so much for your help!

I'm going to close this, because it appears to not be a problem with shelljs. Feel free to reopen if needed.
